hazelcast/src/main/java/com/hazelcast/partition/PartitionAware.java

```diff
@@ -37,6 +37,15 @@
  * {@link com.hazelcast.core.DistributedObject} also has a notion of the partition key which is of type String
  * to ensure that the same partition as distributed Objects Strings is used for the partition key.
  *
+ * Additionally, it is expected that partition key remains same for a given object throughout its lifetime in the
+ * hazelcast storage and the key itself is produced from the value of the Object (following similar semantics to
+ * equals and hashcode). For example if at the time of Object insertion getPartitionKey() returns "test"
+ * as the partition key, it is expected that this Object instance or equivalent Object instances will always
+ * return "test" as the partition key. Returning different partition keys at the time of insertion and at the time
+ * of querying (for a given otherwise equivalent Object), while technically possible, will likely result in inability
```

Comment:
>while technically possible

such behavior breaks the contract - period. We do not need to consider what would happen if the contract was broken as nothing good will happen (similar to breaking hashCode contract). Instead focus on specifying the contract.

Saying "technically possible" might suggest that it is somehow allowed. In our case it is never a good idea to break the contract.